kyuubi-server/src/main/scala/org/apache/kyuubi/engine/ApplicationOperation.scala

```diff
@@ -91,9 +91,17 @@ object ApplicationState extends Enumeration {
   type ApplicationState = Value
   val PENDING, RUNNING, FINISHED, KILLED, FAILED, ZOMBIE, NOT_FOUND, UNKNOWN = Value
 
-  def isFailed(state: ApplicationState): Boolean = state match {
+  def isFailed(
+      state: ApplicationState,
+      appOperation: Option[ApplicationOperation]): Boolean = state match {
     case FAILED => true
     case KILLED => true
+    case NOT_FOUND => appOperation match {
+        // For YARN and Kubernetes operations, if the application is not found, treat it as failed
+        // to prevent mistakenly set unsuccessful applications to the finished state.
+        case Some(_: YarnApplicationOperation) | Some(_: KubernetesApplicationOperation) => true
```

Comment:
Instead of judging the class, how about exposing a method, i.e. `supportPersistedAppState`

Reply:
Sounds good.